File: 프로그래머스/lv3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    g = {i: [] for i in set(genres)}
    g_n = {i: 0 for i in set(genres)}
    
    for i in range(len(plays)):
        g[genres[i]].append([i, plays[i]])
        g_n[genres[i]] += plays[i]
        
    g_n = sorted(g_n.items(), key=lambda x: x [1], reverse = True)

    for a, b in g_n:
        tmp = g[a]
        tmp.sort(key = lambda x: (-x[1], x[0]))
        for i in tmp[:2]:
            answer.append(i[0])
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const map = new Map();

    for (let i = 0; i < genres.length; i++) {
        if (!map.has(genres[i])) {
            map.set(genres[i], []);
        }
        map.get(genres[i]).push([plays[i], i]);
    }

    const sorted = [...map.entries()]
    .sort((a, b) => {
        const totalA = a[1].reduce((sum, [play]) => sum + play, 0);
        const totalB = b[1].reduce((sum, [play]) => sum + play, 0);
        return totalB - totalA;
    });
    
    const answer = [];
    
    for(const [genres, songs] of sorted){
        const sortedSongs = songs.sort((a, b) => {
            if(b[0] != a[0]) return b[0] - a[0];
            return a[1] - b[1];
        });
        
        answer.push(sortedSongs[0][1]);
        if (sortedSongs.length > 1) {
            answer.push(sortedSongs[1][1]);
        }
    }
    return answer;
}
